sql/core/src/main/scala/org/apache/spark/sql/scripting/SqlScriptingExecutionNode.scala

```diff
@@ -155,3 +157,98 @@ abstract class CompoundNestedStatementIteratorExec(collection: Seq[CompoundState
  */
 class CompoundBodyExec(statements: Seq[CompoundStatementExec])
   extends CompoundNestedStatementIteratorExec(statements)
+
+/**
+ * Executable node for IfElseStatement.
+ * @param conditions Collection of executable conditions. First condition corresponds to IF clause,
+ *                   while others (if any) correspond to following ELSE IF clauses.
+ * @param bodies  Collection of executable bodies.
+ * @param booleanEvaluator Evaluator for Boolean conditions.
+ */
+class IfElseStatementExec(
+    conditions: Seq[SingleStatementExec],
+    bodies: Seq[CompoundBodyExec],
+    booleanEvaluator: StatementBooleanEvaluator) extends NonLeafStatementExec {
+  private object IfElseState extends Enumeration {
+    val Condition, Body = Value
+  }
+
+  private var state = IfElseState.Condition
+  private var curr: Option[CompoundStatementExec] = Some(conditions.head)
+
+  private var clauseIdx: Int = 0
+  private val conditionsCount: Int = conditions.length
+  private val bodiesCount: Int = bodies.length
+  assert(conditionsCount == bodiesCount || conditionsCount + 1 == bodiesCount)
+
+  private lazy val treeIterator: Iterator[CompoundStatementExec] =
```

Comment:
so the conditions are part of the resultsets of the if-else statement?

Reply:
yes, the conditions are also considered as the statements that are iterated through.
however, during the condition evaluation, it is marked as `isExecuted` - this way, we allow this decision (whether or not to be included in the result set) to be made outside of the iteration logic - @miland-db is introducing this logic in his [execution PR](https://github.com/apache/spark/pull/47403).
there are multiple reasons for such approach, but mainly we want to iterate through conditions in the same way as with other statements, for the sake of handling exceptions, being able to stop at the condition during debugging, etc.

Reply:
Is it the same in other databases?

Reply:
Not sure what exactly do you mean - do you think from implementation or user perspective?

What I mentioned in the comment above is nothing visible from the user perspective because they cannot explicitly drive the interpreter execution. For more details, you can check the execution PR linked in the previous comment - `SqlScriptingInterpreter` will have `execute()` function, in which we determine whether or not we are collecting results for specific statement. This is where we decide not to take results from conditions.
This PR also sorts out the naming, so `isExecuted` flag that we use here will be changed to more meaningful `shouldCollectResults` flag.

Reply:
I mean from user's perspective within JDBC protocol. Should users be able to see the if conditions as part of the result sets?

Reply:
I'm not sure, will check soon.

Reply:
I've read through JDBC/ODBC API specifications and it seems like my previous understanding was correct - they propose to return result sets using cursors, where you actually specify in the Script/Stored Procedure what are you actually returning, thus it's impossible to return IF conditions in the result sets.

On another hand, MySql has "Unbound SELECT statements", which is really similar to what we want to achieve here - any standalone SQL statement is returned as a part of the result set, but that also means that IF conditions are never returned.

This is the understanding from discussions with Serge as well, we want to return results only for standalone statements.

Reply:
OK so how we evaluate IF conditions is implementation detail.